kybra/compiler/kybra_post_install/src/main.rs

```diff
@@ -0,0 +1,402 @@
+
+    let canister_id = String::from_utf8_lossy(&canister_id_output.stdout)
+        .trim()
+        .to_string();
+
+    Ok(canister_id)
+}
+
+fn get_canister_already_its_own_controller(
+    canister_name: &str,
+    dfx_network: &str,
+    canister_id: &str,
+) -> Result<bool, String> {
+    let current_controllers_output = Command::new("dfx")
+        .arg("canister")
+        .arg("info")
+        .arg("--network")
+        .arg(dfx_network)
+        .arg(canister_name)
+        .output()
+        .map_err(|_| create_error_string("Failed to execute the dfx canister info command"))?;
+
+    if !current_controllers_output.status.success() {
+        return Err(create_error_string(&String::from_utf8_lossy(
+            &current_controllers_output.stderr,
+        )));
+    }
+
+    let canister_already_its_own_controller =
+        String::from_utf8_lossy(&current_controllers_output.stdout)
+            .to_string()
+            .contains(&canister_id);
+
+    Ok(canister_already_its_own_controller)
+}
+
+fn add_controller(canister_name: &str, dfx_network: &str, canister_id: &str) -> Result<(), String> {
+    let add_controller_output = Command::new("dfx")
+        .arg("canister")
+        .arg("update-settings")
+        .arg("--network")
+        .arg(dfx_network)
+        .arg("--add-controller")
+        .arg(&canister_id)
+        .arg(canister_name)
+        .output()
+        .map_err(|_| create_error_string("Failed to execute the dfx command"))?;
+
+    if !add_controller_output.status.success() {
+        return Err(create_error_string(&String::from_utf8_lossy(
+            &add_controller_output.stderr,
+        )));
+    }
+
+    Ok(())
+}
+
+fn install_app_canister(canister_name: &str, dfx_network: &str) -> Result<(), String> {
+    println!("\nFinalizing...\n");
+
+    let install_output = Command::new("dfx")
+        .arg("canister")
+        .arg("call")
+        .arg("--network")
+        .arg(dfx_network)
+        .arg(canister_name)
+        .arg("install_wasm")
+        .output()
+        .map_err(|_| create_error_string("Failed to execute the dfx command"))?;
+
+    if !install_output.status.success() {
+        handle_install_app_canister_failure(&install_output)?;
+    }
+
+    Ok(())
+}
+
+fn handle_install_app_canister_failure(
+    install_output: &std::process::Output,
+) -> Result<(), String> {
+    let error_message = String::from_utf8_lossy(&install_output.stderr);
+
+    if !error_message.contains("did not reply to the call")
+        && !error_message.contains("function invocation does not match its signature")
+    {
+        return Err(create_error_string(&String::from_utf8_lossy(
+            &install_output.stderr,
+        )));
+    }
+
+    Ok(())
+}
+
+fn remove_permissions(
+    canister_name: &str,
+    dfx_network: &str,
+    canister_id: &str,
+    canister_already_its_own_controller: bool,
+) -> Result<(), String> {
+    if canister_already_its_own_controller {
```

Comment:
I wonder if it makes more sense to do this check internally than to have the caller know if they're already a controller... It would increase the complexity of this function but it makes more sense to me.

Reply:
Rename